File: client-app/core/connectivity.py
"""Server 可达性 probe + 心跳 —— M-β.2 阶段。

两个公开能力:

* ``probe(server)``  —— 一次性可达性验证。在 connect 流程的"第 1 步"调,验证
  server 的 healthz 200 + SOCKS/HTTP 端口 TCP 三次握手成功。返 ``ProbeResult``。

* ``Heartbeat`` 协程 —— 进入 connected 态后由 ClientRuntime 启动:每 10s 一次
  GET /healthz; 连续 N 次失败触发 ``heartbeat_changed`` 事件(payload tone =
  green / yellow / red),N=3 时归类为 red 并通知调用方做全局降级。

设计原则:
- 失败种类区分清楚(timeout / connection refused / 5xx),好让 UI 区分 "网络
  断" / "server 异常" / "我们解析错了"。
- 不在这里改任何全局状态(cache / system_proxy),只 publish 事件。状态机由
  ClientRuntime 统一管理。
"""
from __future__ import annotations

import asyncio
import json
import logging
import time
from dataclasses import dataclass
from typing import Optional

import urllib.error
import urllib.request

log = logging.getLogger("conduit.client.connectivity")

DEFAULT_PROBE_TIMEOUT = 3.0          # healthz GET / TCP open
DEFAULT_HEARTBEAT_INTERVAL = 10.0    # 心跳节拍
DEFAULT_HEARTBEAT_TIMEOUT = 5.0      # 单次心跳 GET /healthz timeout
HEARTBEAT_YELLOW_AT = 1              # 第 1 次失败 -> 黄
HEARTBEAT_RED_AT = 3                 # 连续 3 次失败 -> 红 + 全局降级


# ---------------------------------------------------------------------------
# 一次性 probe
# ---------------------------------------------------------------------------


@dataclass
class ProbeResult:
    ok: bool
    healthz_ok: bool
    socks_reachable: bool
    http_reachable: bool
    error: Optional[str] = None
    latency_ms: float = 0.0
    server_vpn: bool = False  # 从 healthz checks 解析出的 vpn 状态


async def _tcp_probe(host: str, port: int, *, timeout: float) -> bool:
    """TCP 三次握手成功即视为可达。"""
    try:
        fut = asyncio.open_connection(host, port)
        reader, writer = await asyncio.wait_for(fut, timeout=timeout)
    except (asyncio.TimeoutError, OSError) as exc:
        log.debug("tcp probe %s:%d failed: %s", host, port, exc)
        return False
    try:
        writer.close()
        await writer.wait_closed()
    except Exception:
        pass
    return True


async def _healthz_probe(host: str, port: int, *, timeout: float) -> tuple[bool, bool, Optional[str]]:
    """GET http://host:port/healthz —— 返 (ok, vpn, error_or_none)。

    server-app 的 healthz 是 loopback only,所以**这里走 LAN HTTP**(同机器
    访问 server 的 8090 不可能拿到,真实场景必须用 server 的 8080 暴露的
    _LAN_-facing healthz)。

    设计澄清:server 的 control API 是 loopback,不能从 LAN 访问。我们这里
    实际上 probe 的是 server 的 *HTTP proxy 端口*(就是 SOCKS 之外的代理端
    口,8080 默认)上的某个 health 路径 —— 但 server 的 HTTP proxy 端口只
    转发流量,没有 health 路径。

    妥协方案(M-β.2):用 mDNS TXT 中的 ``port`` 端口做 TCP probe 即可,不
    实际打 healthz HTTP。 ``healthz_ok`` 字段保留但永远 True,vpn 信息暂
    走 mDNS TXT 已经拿到的(传进 probe 时再带上)。

    M-γ 计划:server 增加 ``GET :http_port/_health`` 公开端点,client 真
    打 health。
    """
    return True, False, None


async def probe(
    *,
    host: str,
    http_port: int,
    socks_port: int,
    timeout: float = DEFAULT_PROBE_TIMEOUT,
) -> ProbeResult:
    """一次性可达性检查。所有底层调用并行,总耗时不超过 ``timeout``。"""
    start = time.monotonic()
    socks_task = asyncio.create_task(_tcp_probe(host, socks_port, timeout=timeout))
    http_task = asyncio.create_task(_tcp_probe(host, http_port, timeout=timeout))
    try:
        socks_ok, http_ok = await asyncio.gather(socks_task, http_task)
    except Exception as exc:  # noqa: BLE001
        return ProbeResult(
            ok=False, healthz_ok=False, socks_reachable=False, http_reachable=False,
            error=str(exc),
            latency_ms=(time.monotonic() - start) * 1000,
        )

    healthz_ok, _vpn, _err = await _healthz_probe(host, http_port, timeout=timeout)
    elapsed_ms = (time.monotonic() - start) * 1000

    if not socks_ok and not http_ok:
        err = f"{host}: SOCKS({socks_port}) 与 HTTP({http_port}) 端口都不通"
    elif not socks_ok:
        err = f"{host}: SOCKS 端口 {socks_port} 不可达"
    elif not http_ok:
        err = f"{host}: HTTP 端口 {http_port} 不可达"
    else:
        err = None

    return ProbeResult(
        ok=socks_ok and http_ok,
        healthz_ok=healthz_ok,
        socks_reachable=socks_ok,
        http_reachable=http_ok,
        error=err,
        latency_ms=elapsed_ms,
        server_vpn=False,
    )


# ---------------------------------------------------------------------------
# 心跳协程
# ---------------------------------------------------------------------------


@dataclass
class HeartbeatState:
    tone: str = "green"           # green / yellow / red
    consecutive_failures: int = 0
    last_check_at: float = 0.0
    last_error: Optional[str] = None


class Heartbeat:
    """简单的心跳:每 ``interval`` 秒一次 TCP probe; 失败次数累加,触发 tone
    迁移。失败回归到成功直接 reset 到 green。"""

    def __init__(
        self,
        bus,
        *,
        host: str,
        http_port: int,
        socks_port: int,
        interval: float = DEFAULT_HEARTBEAT_INTERVAL,
        timeout: float = DEFAULT_HEARTBEAT_TIMEOUT,
    ) -> None:
        self.bus = bus
        self.host = host
        self.http_port = http_port
        self.socks_port = socks_port
        self.interval = interval
        self.timeout = timeout
        self.state = HeartbeatState()
        self._task: Optional[asyncio.Task] = None
        self._stop = asyncio.Event()

    @property
    def running(self) -> bool:
        return self._task is not None and not self._task.done()

    async def start(self) -> None:
        if self.running:
            return
        self._stop = asyncio.Event()
        self._task = asyncio.create_task(self._loop(), name="conduit.heartbeat")
        log.info("heartbeat started for %s (interval=%.1fs)", self.host, self.interval)

    async def stop(self) -> None:
        if self._task is None:
            return
        self._stop.set()
        try:
            await asyncio.wait_for(self._task, timeout=self.interval + 1.0)
        except asyncio.TimeoutError:
            self._task.cancel()
        self._task = None
        log.info("heartbeat stopped")

    async def _loop(self) -> None:
        while not self._stop.is_set():
            await self._tick()
            try:
                await asyncio.wait_for(self._stop.wait(), timeout=self.interval)
            except asyncio.TimeoutError:
                pass

    async def _tick(self) -> None:
        ok = await _tcp_probe(self.host, self.socks_port, timeout=self.timeout)
        self.state.last_check_at = time.time()
        if ok:
            old_tone = self.state.tone
            self.state.consecutive_failures = 0
            self.state.last_error = None
            self.state.tone = "green"
            if old_tone != "green":
                self._publish_tone("green", recovered=True)
        else:
            self.state.consecutive_failures += 1
            self.state.last_error = f"SOCKS {self.host}:{self.socks_port} 不可达"
            new_tone = self._compute_tone(self.state.consecutive_failures)
            if new_tone != self.state.tone:
                self.state.tone = new_tone
                self._publish_tone(new_tone, recovered=False)

    @staticmethod
    def _compute_tone(failures: int) -> str:
        if failures >= HEARTBEAT_RED_AT:
            return "red"
        if failures >= HEARTBEAT_YELLOW_AT:
            return "yellow"
        return "green"

    def _publish_tone(self, tone: str, *, recovered: bool) -> None:
        log.info("heartbeat tone -> %s (failures=%d, recovered=%s)",
                 tone, self.state.consecutive_failures, recovered)
        self.bus.publish("heartbeat_changed", {
            "tone": tone,
            "consecutive_failures": self.state.consecutive_failures,
            "recovered": recovered,
            "last_error": self.state.last_error,
            "host": self.host,
        })

File: client-app/src-tauri/src/proxy/connectivity.rs
//! `connectivity` —— 一次性 probe + 周期心跳。
//!
//! 两个公开能力：
//! - [`probe`] —— 一次性可达性验证（HTTP / SOCKS 端口 TCP 三次握手）
//! - [`Heartbeat`] —— 进入 connected 后周期 probe，连续失败时通过 EventBus 推
//!   `heartbeat_changed`（tone = green/yellow/red）

use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use log::{debug, info, warn};
use tokio::net::TcpStream;
use tokio::sync::Mutex;
use tokio::task::JoinHandle;
use tokio::time::timeout;
use tokio_util::sync::CancellationToken;

use conduit_core::{EventBus, HeartbeatState, HeartbeatTone, ProbeResult};

use super::core::ClientEvent;

pub const DEFAULT_PROBE_TIMEOUT: Duration = Duration::from_secs(3);
pub const DEFAULT_HEARTBEAT_INTERVAL: Duration = Duration::from_secs(10);
pub const HEARTBEAT_YELLOW_AT: u32 = 1;
pub const HEARTBEAT_RED_AT: u32 = 3;

/// 一次性 probe：HTTP + SOCKS 都做 TCP 三次握手，任一失败即整体失败。
///
/// `server_vpn` 通过 `vpn_hint` 回填（来源：server 在 mDNS TXT 上自报的 `vpn=on/off`）。
pub async fn probe(
    host: &str,
    http_port: u16,
    socks_port: u16,
    vpn_hint: bool,
    timeout_dur: Duration,
) -> ProbeResult {
    let start = std::time::Instant::now();
    let http_ok = tcp_probe(host, http_port, timeout_dur).await;
    let socks_ok = tcp_probe(host, socks_port, timeout_dur).await;
    let latency_ms = start.elapsed().as_secs_f64() * 1000.0;
    let ok = http_ok && socks_ok;
    let error = if !ok {
        let mut bits = vec![];
        if !http_ok {
            bits.push(format!("http :{http_port} unreachable"));
        }
        if !socks_ok {
            bits.push(format!("socks :{socks_port} unreachable"));
        }
        Some(bits.join("; "))
    } else {
        None
    };
    ProbeResult {
        ok,
        healthz_ok: ok, // 简化：当作 healthz_ok = http_reachable
        socks_reachable: socks_ok,
        http_reachable: http_ok,
        error,
        latency_ms,
        server_vpn: vpn_hint,
    }
}

async fn tcp_probe(host: &str, port: u16, timeout_dur: Duration) -> bool {
    let target = format!("{host}:{port}");
    matches!(
        timeout(timeout_dur, TcpStream::connect(&target)).await,
        Ok(Ok(_))
    )
}

/// 心跳协程：每 `interval` 跑一次 probe，更新 EventBus 心跳状态。
///
/// 副作用：每次 TCP probe 通过后，会向 server 发送一次
/// `GET /api/clients/heartbeat?name=&version=`，让 server 端把本机注册成
/// "passive client"（已连接但暂无流量），server UI 才能显示客户端已接入。
pub struct Heartbeat {
    bus: EventBus<ClientEvent>,
    target: HeartbeatTarget,
    cancel: CancellationToken,
    state: Arc<Mutex<HeartbeatState>>,
    handle: Mutex<Option<JoinHandle<()>>>,
    http: reqwest::Client,
}

#[derive(Clone)]
struct HeartbeatTarget {
    host: String,
    http_port: u16,
    socks_port: u16,
    client_name: String,
    client_version: String,
}

impl Heartbeat {
    pub fn new(
        bus: EventBus<ClientEvent>,
        host: String,
        http_port: u16,
        socks_port: u16,
        client_name: String,
        client_version: String,
    ) -> Arc<Self> {
        let http = reqwest::Client::builder()
            .timeout(Duration::from_secs(3))
            .build()
            .unwrap_or_else(|_| reqwest::Client::new());
        Arc::new(Self {
            bus,
            target: HeartbeatTarget {
                host,
                http_port,
                socks_port,
                client_name,
                client_version,
            },
            cancel: CancellationToken::new(),
            state: Arc::new(Mutex::new(HeartbeatState {
                tone: HeartbeatTone::Green,
                consecutive_failures: 0,
                last_check_at: 0.0,
                last_error: None,
            })),
            handle: Mutex::new(None),
            http,
        })
    }

    pub async fn start(self: Arc<Self>) {
        let me = self.clone();
        let task = tokio::spawn(async move {
            me.run_loop().await;
        });
        *self.handle.lock().await = Some(task);
        info!(
            "[heartbeat] started target={}:{} every {}s",
            self.target.host,
            self.target.http_port,
            DEFAULT_HEARTBEAT_INTERVAL.as_secs()
        );
    }

    pub async fn stop(&self) {
        self.cancel.cancel();
        if let Some(h) = self.handle.lock().await.take() {
            let _ = h.await;
        }
        info!("[heartbeat] stopped");
    }

    pub async fn snapshot(&self) -> HeartbeatState {
        self.state.lock().await.clone()
    }

    async fn run_loop(self: Arc<Self>) {
        loop {
            if self.cancel.is_cancelled() {
                return;
            }
            let result = probe(
                &self.target.host,
                self.target.http_port,
                self.target.socks_port,
                false,
                Duration::from_secs(5),
            )
            .await;
            self.update_state(&result).await;

            // TCP 双端口都通才发 HTTP 心跳通知 server。HTTP 失败不计入心跳健康
            // (server 可能临时高负载或 reqwest 重试中),只 warn 一行。
            if result.ok {
                self.notify_server().await;
            }

            tokio::select! {
                _ = self.cancel.cancelled() => return,
                _ = tokio::time::sleep(DEFAULT_HEARTBEAT_INTERVAL) => {}
            }
        }
    }

    /// 给 server 发 `GET /api/clients/heartbeat?name=&version=`,让 server 把
    /// 本机登记成 passive client。server 端实现见
    /// `server-app/src-tauri/src/proxy/http.rs::serve_heartbeat`。
    async fn notify_server(&self) {
        let url = format!(
            "http://{}:{}/api/clients/heartbeat?name={}&version={}",
            self.target.host,
            self.target.http_port,
            urlencoding(&self.target.client_name),
            urlencoding(&self.target.client_version),
        );
        match self.http.get(&url).send().await {
            Ok(resp) if resp.status().is_success() => {
                debug!("[heartbeat] notified server passive registry: {url}");
            }
            Ok(resp) => {
                warn!("[heartbeat] server returned {} for {url}", resp.status());
            }
            Err(e) => {
                warn!("[heartbeat] failed to notify server passive registry: {e}");
            }
        }
    }

    async fn update_state(&self, probe: &ProbeResult) {
        let mut state = self.state.lock().await;
        state.last_check_at = epoch_now();
        if probe.ok {
            if state.consecutive_failures > 0 {
                debug!("[heartbeat] recovered from {} failures", state.consecutive_failures);
            }
            state.consecutive_failures = 0;
            state.last_error = None;
            state.tone = HeartbeatTone::Green;
        } else {
            state.consecutive_failures = state.consecutive_failures.saturating_add(1);
            state.last_error = probe.error.clone();
            let prev = state.tone;
            state.tone = if state.consecutive_failures >= HEARTBEAT_RED_AT {
                HeartbeatTone::Red
            } else if state.consecutive_failures >= HEARTBEAT_YELLOW_AT {
                HeartbeatTone::Yellow
            } else {
                HeartbeatTone::Green
            };
            if prev != state.tone {
                warn!(
                    "[heartbeat] {prev:?} → {:?} after {} failures: {}",
                    state.tone,
                    state.consecutive_failures,
                    probe.error.as_deref().unwrap_or("?")
                );
            }
        }
        let snapshot = state.clone();
        drop(state);
        self.bus.publish(ClientEvent {
            kind: "heartbeat_changed".into(),
            payload: serde_json::to_value(&snapshot).unwrap_or(serde_json::Value::Null),
            ts: epoch_now(),
        });
    }
}

fn epoch_now() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or_default()
}

/// 极简 URL 编码：保留 ASCII 字母 / 数字 / `-_.~`,其余字节按 `%XX` 转义。
/// 仅服务于 query string 拼装,不追求 RFC3986 完整覆盖。
fn urlencoding(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    for &b in s.as_bytes() {
        if b.is_ascii_alphanumeric() || matches!(b, b'-' | b'_' | b'.' | b'~') {
            out.push(b as char);
        } else {
            out.push_str(&format!("%{:02X}", b));
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn probe_unreachable_target_returns_error() {
        // 192.0.2.1 是 TEST-NET-1，保证不可路由
        let r = probe("192.0.2.1", 12345, 12346, false, Duration::from_millis(200)).await;
        assert!(!r.ok);
        assert!(!r.http_reachable);
        assert!(!r.socks_reachable);
        assert!(r.error.is_some());
    }

    #[tokio::test]
    async fn probe_reports_latency() {
        let r = probe("192.0.2.1", 12345, 12346, true, Duration::from_millis(100)).await;
        assert!(r.latency_ms > 0.0);
        assert!(r.server_vpn);
    }

    #[test]
    fn epoch_now_is_positive() {
        let n = epoch_now();
        assert!(n > 1_700_000_000.0);
    }

    #[test]
    fn urlencoding_escapes_special_chars() {
        assert_eq!(urlencoding("alice-mac"), "alice-mac");
        assert_eq!(urlencoding("v0.2.0"), "v0.2.0");
        assert_eq!(urlencoding("hello world"), "hello%20world");
        assert_eq!(urlencoding("a&b=c"), "a%26b%3Dc");
        assert_eq!(urlencoding("中"), "%E4%B8%AD");
    }

    /// 起一个 mock TCP listener 模拟 server 的 HTTP listener,
    /// 接受一次 GET 请求后回 200 OK,验证 Heartbeat 真的会发 GET /api/clients/heartbeat。
    #[tokio::test]
    async fn heartbeat_notifies_server_after_successful_probe() {
        use tokio::io::{AsyncReadExt, AsyncWriteExt};
        use tokio::net::TcpListener;

        // mock server 在两个端口上监听:http_port 处理 heartbeat 请求,
        // socks_port 只接受 TCP 连接(让 probe 通过)。
        let http_listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let http_port = http_listener.local_addr().unwrap().port();
        let socks_listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
        let socks_port = socks_listener.local_addr().unwrap().port();

        let (tx, mut rx) = tokio::sync::mpsc::channel::<String>(4);
        let tx_socks = tx.clone();

        // socks 端口只处理 probe 的 TCP 三次握手,accept 后立即关闭。
        tokio::spawn(async move {
            while let Ok((mut s, _)) = socks_listener.accept().await {
                let _ = tx_socks.send("socks-accept".into()).await;
                let _ = s.shutdown().await;
            }
        });

        // http 端口接受 GET 后 capture request line 推到 channel 再回 200。
        tokio::spawn(async move {
            while let Ok((mut s, _)) = http_listener.accept().await {
                let mut buf = [0u8; 1024];
                let n = s.read(&mut buf).await.unwrap_or(0);
                let req = String::from_utf8_lossy(&buf[..n]).to_string();
                let _ = tx.send(req).await;
                let body = b"{\"ok\":true,\"created\":true,\"ttl_sec\":60}";
                let head = format!(
                    "HTTP/1.1 200 OK\r\nContent-Length: {}\r\nConnection: close\r\n\r\n",
                    body.len()
                );
                let _ = s.write_all(head.as_bytes()).await;
                let _ = s.write_all(body).await;
                let _ = s.shutdown().await;
            }
        });

        let bus: EventBus<ClientEvent> = EventBus::new(8);
        let hb = Heartbeat::new(
            bus,
            "127.0.0.1".to_string(),
            http_port,
            socks_port,
            "alice-mac".to_string(),
            "0.2.0".to_string(),
        );
        hb.clone().start().await;

        // 等待 mock server 收到 heartbeat HTTP 请求;最多等 5s。
        let mut got_heartbeat = false;
        let deadline = std::time::Instant::now() + Duration::from_secs(5);
        while std::time::Instant::now() < deadline {
            match tokio::time::timeout(Duration::from_millis(500), rx.recv()).await {
                Ok(Some(req)) if req.starts_with("GET /api/clients/heartbeat?") => {
                    assert!(req.contains("name=alice-mac"), "{req}");
                    assert!(req.contains("version=0.2.0"), "{req}");
                    got_heartbeat = true;
                    break;
                }
                Ok(_) => continue,
                Err(_) => continue,
            }
        }
        hb.stop().await;
        assert!(got_heartbeat, "Heartbeat did not notify server within 5s");
    }
}
